File: backend/agents/orchestrator.py
import logging
from datetime import datetime, timedelta, timezone
from typing import Any, Optional

from models.request_models import IntentResult, NearbyPlace, WeatherResponse
from services.agent_store import AgentStore
from services.location_service import LocationService, DISTRICT_COORDINATES
from services.weather_service import WeatherService

logger = logging.getLogger("navix.orchestrator")


class TripOrchestrator:
    """Coordinates weather and tourism agents for trip-planning queries.

    When a user says something like "I want to go to Colombo tomorrow", this
    orchestrator:
      1. Detects the travel_date and location from intent entities.
      2. Fetches the weather forecast for that location + date.
      3. If the forecast is bad (rain), finds alternative locations with
         better weather and recommends them.
    """

    def __init__(
        self,
        weather_service: WeatherService,
        location_service: LocationService,
        agent_store: AgentStore | None = None,
    ) -> None:
        self.weather = weather_service
        self.locations = location_service
        self.agent_store = agent_store

    async def run(
        self,
        query: str,
        intent: IntentResult,
        nearby: list[NearbyPlace],
    ) -> Optional[dict[str, Any]]:
        """Run orchestration. Returns context dict or None if not applicable."""
        travel_date = self._extract_travel_date(intent)
        location_name = self._extract_location(intent)

        if not travel_date or not location_name:
            return None

        # Resolve coordinates for the target location
        target_coords = self._resolve_coordinates(location_name, nearby)
        if not target_coords:
            return None

        # Fetch forecast for the target location
        target_forecast = await self._safe_forecast(
            target_coords["lat"], target_coords["lng"], travel_date
        )
        if not target_forecast:
            return None

        result: dict[str, Any] = {
            "trip_planning": True,
            "target_location": location_name,
            "travel_date": travel_date,
            "target_forecast": {
                "temperature": target_forecast.temperature,
                "humidity": target_forecast.humidity,
                "condition": target_forecast.condition,
                "rain": target_forecast.rain,
                "safety_hints": target_forecast.safety_hints,
            },
            "weather_is_bad": target_forecast.rain,
            "alternatives": [],
        }

        # If weather is bad, find alternative locations with better weather
        if target_forecast.rain:
            alternatives = await self._find_sunny_alternatives(
                location_name, travel_date
            )
            result["alternatives"] = alternatives

        if self.agent_store:
            self.agent_store.log(
                "trip_orchestrator",
                query,
                {
                    "location": location_name,
                    "travel_date": travel_date,
                    "target_coords": target_coords,
                },
                result,
                status="success",
            )

        return result

    def _extract_travel_date(self, intent: IntentResult) -> Optional[str]:
        """Extract and normalize the travel date from intent entities."""
        entities = intent.entities or {}
        raw_date = entities.get("travel_date")
        if not raw_date:
            return None

        raw = str(raw_date).lower().strip()
        now = datetime.now(timezone.utc)

        if raw == "today":
            return now.strftime("%Y-%m-%d")
        elif raw == "tomorrow":
            return (now + timedelta(days=1)).strftime("%Y-%m-%d")
        elif raw in ("this weekend",):
            # Find next Saturday
            days_ahead = 5 - now.weekday()  # Saturday = 5
            if days_ahead <= 0:
                days_ahead += 7
            return (now + timedelta(days=days_ahead)).strftime("%Y-%m-%d")
        elif raw in ("next week",):
            return (now + timedelta(days=7)).strftime("%Y-%m-%d")
        else:
            # Try to parse as ISO date
            try:
                parsed = datetime.fromisoformat(raw)
                return parsed.strftime("%Y-%m-%d")
            except ValueError:
                # Default to tomorrow for any unrecognized date reference
                return (now + timedelta(days=1)).strftime("%Y-%m-%d")

    def _extract_location(self, intent: IntentResult) -> Optional[str]:
        """Extract a location name from intent entities."""
        entities = intent.entities or {}
        for key in ("location", "district", "city", "place", "destination"):
            value = entities.get(key)
            if isinstance(value, str) and len(value.strip()) > 1:
                return value.strip()
        return None

    def _resolve_coordinates(
        self, location_name: str, nearby: list[NearbyPlace]
    ) -> Optional[dict[str, float]]:
        """Get coordinates for a location name."""
        # Check district coordinates lookup table
        key = location_name.lower().strip()
        if key in DISTRICT_COORDINATES:
            return DISTRICT_COORDINATES[key]

        # Check nearby places for a name match
        for place in nearby:
            if location_name.lower() in place.name.lower():
                return place.coordinates

        # Try the location service
        results = self.locations.search_by_name(location_name, limit=1)
        if results:
            return results[0].coordinates

        return None

    async def _safe_forecast(
        self, lat: float, lng: float, date_str: str
    ) -> Optional[WeatherResponse]:
        """Fetch forecast, returning None on any error."""
        try:
            return await self.weather.get_forecast(lat, lng, date_str)
        except Exception as exc:
            logger.warning("Forecast fetch failed: %s", exc)
            return None

    async def _find_sunny_alternatives(
        self, exclude_location: str, travel_date: str, max_alternatives: int = 3
    ) -> list[dict[str, Any]]:
        """Find locations in different areas that have good weather on the target date."""
        exclude_key = exclude_location.lower().strip()

        # Pick candidate districts that are geographically spread across Sri Lanka
        candidates = [
            ("Kandy", DISTRICT_COORDINATES.get("kandy")),
            ("Galle", DISTRICT_COORDINATES.get("galle")),
            ("Sigiriya", DISTRICT_COORDINATES.get("sigiriya")),
            ("Ella", DISTRICT_COORDINATES.get("ella")),
            ("Negombo", DISTRICT_COORDINATES.get("negombo")),
            ("Trincomalee", DISTRICT_COORDINATES.get("trincomalee")),
            ("Anuradhapura", DISTRICT_COORDINATES.get("anuradhapura")),
            ("Nuwara Eliya", DISTRICT_COORDINATES.get("nuwara eliya")),
        ]

        alternatives: list[dict[str, Any]] = []
        for name, coords in candidates:
            if name.lower() == exclude_key or not coords:
                continue

            forecast = await self._safe_forecast(
                coords["lat"], coords["lng"], travel_date
            )
            if forecast and not forecast.rain:
                # Get top places in this area
                places = self.locations.search_by_district(name, limit=3)
                place_names = [p.name for p in places[:3]] if places else []

                alternatives.append({
                    "location": name,
                    "forecast": {
                        "temperature": forecast.temperature,
                        "condition": forecast.condition,
                        "rain": forecast.rain,
                    },
                    "top_attractions": place_names,
                })

                if len(alternatives) >= max_alternatives:
                    break

        return alternatives

File: backend/src/agents/orchestrator.ts
/**
 * Trip Orchestrator — coordinates weather forecasts with tourism recommendations.
 *
 * When a user says "I want to go to Colombo tomorrow", this orchestrator:
 *   1. Detects travel_date and location from intent entities
 *   2. Fetches the forecast for that location + date
 *   3. If rain is forecast, finds alternative locations with better weather
 */

import { searchPlacesByDistrict } from "../db";
import { Env, IntentResult, NearbyPlace, WeatherResponse } from "../types";
import { getWeatherForecast } from "../weather";

/** Approximate center coordinates for Sri Lankan districts / cities. */
const DISTRICT_COORDS: Record<string, { lat: number; lng: number }> = {
  colombo:       { lat: 6.9271, lng: 79.8612 },
  kandy:         { lat: 7.2906, lng: 80.6337 },
  galle:         { lat: 6.0535, lng: 80.2210 },
  matara:        { lat: 5.9549, lng: 80.5550 },
  jaffna:        { lat: 9.6615, lng: 80.0255 },
  trincomalee:   { lat: 8.5874, lng: 81.2152 },
  anuradhapura:  { lat: 8.3114, lng: 80.4037 },
  polonnaruwa:   { lat: 7.9403, lng: 81.0003 },
  sigiriya:      { lat: 7.9570, lng: 80.7603 },
  ella:          { lat: 6.8754, lng: 81.0465 },
  negombo:       { lat: 7.2096, lng: 79.8380 },
  bentota:       { lat: 6.4262, lng: 79.9994 },
  "nuwara eliya": { lat: 6.9497, lng: 80.7891 },
  dambulla:      { lat: 7.8568, lng: 80.6491 },
  hikkaduwa:     { lat: 6.1395, lng: 80.1037 },
  unawatuna:     { lat: 6.0107, lng: 80.2494 },
  mirissa:       { lat: 5.9453, lng: 80.4546 },
  tangalle:      { lat: 6.0243, lng: 80.7946 },
  weligama:      { lat: 5.9740, lng: 80.4292 },
  hambantota:    { lat: 6.1429, lng: 81.1212 },
  badulla:       { lat: 6.9934, lng: 81.0550 },
  ratnapura:     { lat: 6.6828, lng: 80.3992 },
  kurunegala:    { lat: 7.4863, lng: 80.3647 },
};

export interface OrchestratorResult {
  trip_planning: true;
  target_location: string;
  travel_date: string;
  target_forecast: {
    temperature: number;
    humidity: number;
    condition: string;
    rain: boolean;
    safety_hints: string[];
  };
  weather_is_bad: boolean;
  alternatives: Array<{
    location: string;
    forecast: { temperature: number; condition: string; rain: boolean };
    top_attractions: string[];
  }>;
}

export async function runTripOrchestrator(
  env: Env,
  query: string,
  intent: IntentResult,
  nearby: NearbyPlace[]
): Promise<Record<string, unknown> | null> {
  const travelDate = extractTravelDate(intent);
  const locationName = extractLocation(intent);

  if (!travelDate || !locationName) return null;

  const coords = resolveCoords(locationName, nearby);
  if (!coords) return null;

  // Fetch forecast for target
  let forecast: WeatherResponse;
  try {
    forecast = await getWeatherForecast(env, coords.lat, coords.lng, travelDate);
  } catch {
    return null;
  }

  const result: OrchestratorResult = {
    trip_planning: true,
    target_location: locationName,
    travel_date: travelDate,
    target_forecast: {
      temperature: forecast.temperature,
      humidity: forecast.humidity,
      condition: forecast.condition,
      rain: forecast.rain,
      safety_hints: forecast.safety_hints,
    },
    weather_is_bad: forecast.rain,
    alternatives: [],
  };

  // If bad weather, find sunny alternatives
  if (forecast.rain) {
    result.alternatives = await findSunnyAlternatives(
      env, locationName, travelDate
    );
  }

  return result as unknown as Record<string, unknown>;
}

// ── Helpers ──

function extractTravelDate(intent: IntentResult): string | null {
  const entities = intent.entities ?? {};
  const raw = entities.travel_date;
  if (!raw || typeof raw !== "string") return null;

  const lower = raw.toLowerCase().trim();
  const now = new Date();

  if (lower === "today") {
    return now.toISOString().slice(0, 10);
  }
  if (lower === "tomorrow") {
    return new Date(now.getTime() + 86400000).toISOString().slice(0, 10);
  }
  if (lower === "this weekend") {
    const daysToSat = (6 - now.getDay() + 7) % 7 || 7;
    return new Date(now.getTime() + daysToSat * 86400000).toISOString().slice(0, 10);
  }
  if (lower === "next week") {
    return new Date(now.getTime() + 7 * 86400000).toISOString().slice(0, 10);
  }
  // Try ISO parse
  const d = new Date(raw);
  if (!isNaN(d.getTime())) return d.toISOString().slice(0, 10);

  // Fallback to tomorrow
  return new Date(now.getTime() + 86400000).toISOString().slice(0, 10);
}

function extractLocation(intent: IntentResult): string | null {
  const entities = intent.entities ?? {};
  for (const key of ["location", "district", "city", "place", "destination"]) {
    const v = entities[key];
    if (typeof v === "string" && v.trim().length > 1) return v.trim();
  }
  return null;
}

function resolveCoords(
  name: string,
  nearby: NearbyPlace[]
): { lat: number; lng: number } | null {
  const key = name.toLowerCase().trim();
  if (DISTRICT_COORDS[key]) return DISTRICT_COORDS[key];

  for (const p of nearby) {
    if (p.name.toLowerCase().includes(key)) return p.coordinates;
  }
  return null;
}

async function findSunnyAlternatives(
  env: Env,
  excludeLocation: string,
  travelDate: string,
  maxAlts = 3
): Promise<OrchestratorResult["alternatives"]> {
  const excludeKey = excludeLocation.toLowerCase().trim();
  const candidates = [
    "kandy", "galle", "sigiriya", "ella", "negombo",
    "trincomalee", "anuradhapura", "nuwara eliya",
  ];

  const alts: OrchestratorResult["alternatives"] = [];

  for (const name of candidates) {
    if (name === excludeKey) continue;
    const coords = DISTRICT_COORDS[name];
    if (!coords) continue;

    let forecast: WeatherResponse;
    try {
      forecast = await getWeatherForecast(env, coords.lat, coords.lng, travelDate);
    } catch {
      continue;
    }

    if (!forecast.rain) {
      // Get top attractions from D1 if available
      let topAttractions: string[] = [];
      if (env.DB) {
        try {
          const places = await searchPlacesByDistrict(env.DB, name, 3);
          topAttractions = places.map((p) => p.name);
        } catch {
          topAttractions = [];
        }
      }

      alts.push({
        location: capitalize(name),
        forecast: {
          temperature: forecast.temperature,
          condition: forecast.condition,
          rain: forecast.rain,
        },
        top_attractions: topAttractions,
      });

      if (alts.length >= maxAlts) break;
    }
  }

  return alts;
}

function capitalize(s: string): string {
  return s ? s.charAt(0).toUpperCase() + s.slice(1) : s;
}
